lib/data/carp_backend.dart

```diff
@@ -151,25 +151,36 @@ class CarpBackend {
     return invitations;
   }
 
-  /// Upload the result of an informed consent flow.
-  Future<ConsentDocument?> uploadInformedConsent(
-      RPTaskResult taskResult) async {
-    RPConsentSignatureResult signatureResult =
-        (taskResult.results["consentreviewstepID"] as RPConsentSignatureResult);
-    signatureResult.userID = username;
-    Map<String, dynamic> informedConsent = signatureResult.toJson();
+  /// Upload the result of an informed consent flow. Returns the uploaded
+  /// consent document as retrieved from the server.
+  ///
+  /// Looks for the first instance of a [RPConsentSignatureResult] in [consent]
+  /// and uploads this.
+  Future<ConsentDocument?> uploadInformedConsent(RPTaskResult consent) async {
+    late RPConsentSignatureResult signedConsent;
+    try {
+      signedConsent = consent.results.values.firstWhere(
+        (result) => result is RPConsentSignatureResult,
+      ) as RPConsentSignatureResult;
+    } catch (_) {
+      warning(
+          '$runtimeType - No signed informed consent found to be uploaded.');
+      return null;
+    }
+
+    signedConsent.userID = username;
+    Map<String, dynamic> json = signedConsent.toJson();
 
-    ConsentDocument? document;
+    ConsentDocument? uploadedConsent;
     try {
-      document = await CarpService().createConsentDocument(informedConsent);
+      uploadedConsent = await CarpService().createConsentDocument(json);
       info(
-          'Informed consent document uploaded successfully - id: ${document.id}');
-      bloc.hasInformedConsentBeenAccepted = true;
```

Comment:
I can see that this has been removed, is this no longer needed?

Reply:
Updating `bloc` on both success and failure ^

Reply:
No. This has been moved to the bloc in order to keep the semantics of this method clear (it's only responsibility is to upload and not to mark anything locally in the app's BLoC). This is now handled in the bloc - see

https://github.com/cph-cachet/carp_studies_app/blob/4b2de58f0f4f7c2ef2dbf9a312ef4c50d164b966/lib/blocs/app_bloc.dart#L264